lib/state_machine.js

```diff
@@ -27,99 +27,196 @@ limitations under the License.
+      // some other state. If we allow this, the external event might be lost,
+      // indicating that something has gone truly wrong. In order to ensure 
+      // that this can't happen, we queue up a single external event in a
+      // separate variable, and always prefer it over any internal event to the
+      // same state. This can means that the author of the States in the State
+      // Machine can't always be assured that an internal transition will occur,
+      // and must guard against any transition happening at any time. But this
+      // is really no different than the work an implementor has to do.
+  
+      if (this.nextExternalState_) {
+        throw new Error('There already IS a next state');
+      }
+      if (this.debug_) {
+        this.debug_(`External transition to ${newState.getName()} requested.`);
+      }
+    
+      this.nextExternalState_ = newState;
+      this.resolver_(newState);
+    }
+  
+    /**
+     * Sets the shared context object to the passed-in value.
+     * @param {*} context
+     */
+    setContext(context) {
+      this.context_ = context;
+    }
+  
+    /**
+     * Returns a promise that's resolved when the current transition finishes. If
+     * there is no current transitions, it will resolve when the next transition
+     * requested completes.
+     * @package
+     * @return {!Promise<STATETYPE>}
+     */
+    getTransitionPromise() {
+      return new Promise((resolve, reject) => {
+        this.transitionPromise_.then(resolve, reject);
+      });
+    }
+  
+    /**
+     * When an exception is thrown during a transition, we call this method.
+     * @abstract
+     * @protected
+     * @param {!Error} error
+     */
+    handleError(error) {}
+  
+    /**
+     * Waits for a transition to be requested, then performs the transition,
+     * finally waiting again.
+     * @protected
+     */
+    driveMachine() {
+      this.transitionPromise_.then(newState => {
+        // Tell old state to go away (synchronously).
+        this.state.exit();
+        if (this.nextExternalState_) {
+          // External transitions always win.
+          newState = this.nextExternalState_;
+          this.nextExternalState_ = null;
+        }
+      
+        if (this.debug_) {
+          this.debug_(`Transitioning from ${this.state.getName()} to ${newState.getName()}.`);
+        }
+      
+        // Adopt the new state, now that cleanup is complete.
+        this.state = newState;
+        // Now, we need to enter the new state, and the state may choose to
+        // move to some other state automatically (over time), but some other
+        // force may drive the state machine to a new state!
+      
+        // First, we need to make a new resolver, indicating that we now allow
+        // outside users to drive the state machine again via transitionTo().
+        let resolver, promise = new Promise(r => resolver = r);
```

Comment:
is new Promise returning 2 values or are both resolver and promise each getting different promises?